classification/feature_generation/feature_utilities.py

```diff
@@ -0,0 +1,373 @@
+        movement feature.
+
+    Args:
+        random_state: a numpy randomstate object.
+        input: the input data as a 2d numpy array.
+        id_: the id of the vessel which made this series.
+        max_time_delta: the maximum time contained in each window.
+        window_size: the size of the window.
+        min_points_for_classification: the minumum number of points in a window for
+            it to be usable.
+
+    Returns:
+      A tuple comprising:
+        1. An numpy array comprising N feature timeslices, of dimension
+            [n, 1, window_size, num_features].
+        2. A numpy array with the timestamps for each feature point, of
+           dimension [n, window_size].
+        3. A numpy array comprising timebounds for each slice, of dimension
+            [n, 2].
+        4. A numpy array with an int64 id for each slice, of dimension [n].
+
+    """
+    slices = []
+    times = input_series[:, 0]
+    for (start_time, end_time) in time_ranges:
+        start_index = np.searchsorted(times, start_time, side='left')
+        end_index = np.searchsorted(times, end_time, side='left')
+        length = end_index - start_index
+
+        # Slice out the time window.
+        cropped = input_series[start_index:end_index]
+
+        # If this window is too long, pick a random subwindow.
+        if (length > window_size):
+            max_offset = length - window_size
+            start_offset = random_state.uniform(max_offset)
+            cropped = cropped[max_offset:max_offset + window_size]
+
+        if len(cropped) >= min_points_for_classification:
+            output_slice = np_pad_repeat_slice(cropped, window_size)
+            time_bounds = np.array([start_time, end_time], dtype=np.int32)
+            without_timestamp = output_slice[:, 1:]
+            timeseries = output_slice[:, 0].astype(np.int32)
+            slices.append(
+                (np.stack([without_timestamp]), timeseries, time_bounds, id_))
+    # TODO: factor out this logic
+    if slices == []:
+        return empty_data(window_size, input_series)
+    return zip(*slices)
+
+
+def np_array_extract_all_fixed_slices(input_series, num_features, id_,
+                                      window_size, shift):
+    slices = []
+    input_length = len(input_series)
+    for end_index in range(input_length, 0, -shift):
+        start_index = end_index - window_size
+        if start_index < 0:
+            continue
+        cropped = input_series[start_index:end_index]
+        start_time = int(cropped[0][0])
+        end_time = int(cropped[-1][0])
+        time_bounds = np.array([start_time, end_time], dtype=np.int32)
+        assert len(cropped) == window_size
+        slices.append(cook_features(cropped, id_))
+    if slices == []:
+        return empty_data(window_size, input_series)
+    return zip(*slices)
+
+
+
+def process_fixed_window_features(random_state, features, id_, 
```

Comment:
Any chance of a comment explaining what this is for?

Reply:
✔️